Move `libstdc++` path into `LOADER_*_DEP_LIBS`

staticfloat opened the pull request:
After adding `libstdc++` probing into the Julia loader [0], we originally made the assumption that the `libstdc++` that is shipped with `julia` would always be co-located with `libjulia.so` [1].  This is not the case when building with `USE_SYSTEM_CSL=1`, however, where we sequester system libraries in `usr/lib/julia`, even at build-time.

The path to `libstdc++.so` has already been getting altered when moving from build-time to install time via `stringreplace` [2], but after further thought, I decided that it would be better to just use the pre-existing `LOADER_*_DEP_LIBS` mechanism to communicate to the loader what the correct relative path to `libstdc++.so` is.  This also allows the single `stringreplace` to update all of our "special" library paths.

[0] https://github.com/JuliaLang/julia/pull/46976
[1] https://github.com/JuliaLang/julia/pull/46976/files#diff-8c5c98f26f3f7aac8905a1074c5bec11a57e9b9c7c556791deac5a3b27cc096fR379 [2] https://github.com/JuliaLang/julia/blob/master/Makefile#L430

Fix #47987.

staticfloat committed:
Move `libstdc++` path into `LOADER_*_DEP_LIBS`

After adding `libstdc++` probing into the Julia loader [0], we
originally made the assumption that the `libstdc++` that is shipped with
`julia` would always be co-located with `libjulia.so` [1].  This is not
the case when building with `USE_SYSTEM_CSL=1`, however, where we
sequester system libraries in `usr/lib/julia`, even at build-time.

The path to `libstdc++.so` has already been getting altered when moving
from build-time to install time via `stringreplace` [2], but after
further thought, I decided that it would be better to just use the
pre-existing `LOADER_*_DEP_LIBS` mechanism to communicate to the loader
what the correct relative path to `libstdc++.so` is.  This also allows
the single `stringreplace` to update all of our "special" library paths.

[0] https://github.com/JuliaLang/julia/pull/46976
[1] https://github.com/JuliaLang/julia/pull/46976/files#diff-8c5c98f26f3f7aac8905a1074c5bec11a57e9b9c7c556791deac5a3b27cc096fR379
[2] https://github.com/JuliaLang/julia/blob/master/Makefile#L430

staticfloat commented:
This should fix #47987

fxcoudert commented:
Fixes the build on Linux for Homebrew. Thanks!

vtjnash commented on cli/loader_lib.c:
```diff
@@ -366,6 +366,50 @@ __attribute__((constructor)) void jl_load_libjulia_internal(void) {
+    // Unpack our special library names.  This is why ordering of library names matters.
+    char * bundled_libstdcxx_path = special_library_names[0];
+    libjulia_internal = load_library(special_library_names[1], lib_dir, 1);
+    void *libjulia_codegen = load_library(special_library_names[2], lib_dir, 0);
```
Why are we loading this early now?

staticfloat replied:
Ahh, that was unintentional.  Thank you, good catch.

staticfloat committed:
Load special libraries in-order

The `DEPS_LIBS` RPATH-substitute mechanism contains a list of paths to
load, and some of these paths are "special", in that they require more
involved loading than simply `load_library()`.  These libraries are
thereby denoted by a `@` prefixing them.

Previously, we made note of these libraries, then loaded them at the end
of the loading loop, but with the addition of `libstdc++` it is now
important to have the order of the libraries (including special
libraries) to be obeyed by the loading loop, so I have inlined special
library handling into the loading loop.  In the future, we may wish to
denote special libraries more explicitly than simply relying on there
being exactly three libraries, with the ordering being mapped to
`libstdc++`, `libjulia-internal`, and `libjulia-codegen`.

staticfloat commented:
I've now pushed a commit that causes the library loading (including special libraries) to obey the order as specified in `DEP_LIBS`.  The current ordering is now:

* `libgcc_s`
* `libm` (usually OpenLibm, but can be system libm if `USE_SYSTEM_LIBM=1`)
* `libstdc++`
* `libjulia-internal`
* `libjulia-codegen`

This ordering is important because `libstdc++` depends on `libgcc_s` and `libm`, so if we are using `USE_SYSTEM_LIBM=1`, we technically should load `libm` first.  Previously, we would probe and load `libstdc++` first, which would bring in `libgcc_s` and `libm` from the system.

vtjnash commented on Make.inc:
```diff
@@ -1465,7 +1465,7 @@ JULIA_SYSIMG_release := $(build_private_libdir)/sys.$(SHLIB_EXT)
 
 define dep_lib_path
-$$($(PYTHON) $(call python_cygpath,$(JULIAHOME)/contrib/relative_path.py) $(1) $(2))
+$(shell $(PYTHON) $(call python_cygpath,$(JULIAHOME)/contrib/relative_path.py) $(1) $(2))
```
:heart:

staticfloat replied:
This slows things down a bit (adds ~60ms to every `make` invocation on my linux desktop) but I think it's worth it for the extra clarity.
